Implement functionality promised in the README

chrismytton opened the pull request:
Following on from #1 and #2 this actually fills in the blanks and makes the RubyGem do what the `README.md` is promising.
# Notes to reviewer

The code for creating an open-uri like response is mostly inspired by/taken from [open-uri itself](https://github.com/ruby/ruby/blob/v2_3_1/lib/open-uri.rb#L410-L424), so I'm reasonably confident that this will produce results identical to open-uri.

chrismytton committed:
Add open-uri-cached development dependency

We don't actually need to use this in the library, but it's useful to
have for generating test data.

chrismytton committed:
Add rake task for pulling in test data

chrismytton committed:
Add some test fixtures

These are some simple requests so that we have some data from
open-uri-cached to work with. This data was created with:

    rake generate_test_data

chrismytton committed:
Add an implementation for the README functionality

This fills in the gem with the actual working functionality for this
library.

Most of the implementation is taken from how open-uri itself does things
[1], which can be a bit hard to follow at times.

[1]: https://github.com/ruby/ruby/blob/v2_3_1/lib/open-uri.rb#L410-L424

chrismytton committed:
Re-generate test fixtures with Ruby 2.1

This will make the tests pass on this Ruby version. They were previously
failing because Ruby < 2.2 doesn't have the `URI::RFC3986_Parser`
constant defined.

In the future we should perhaps add some extra error checking for now
and tell the user that they need to use the same Ruby version that was
used to create the cache files, but for now this will do.

chrismytton committed:
Rename Response class -> CachedResponse

This makes it a bit clearer what the class is actually representing and
makes it a little less jarring seeing `meta_file` being passed into the
constructor.

chrismytton committed:
Rename meta_file variable -> meta_file_path

This makes it clearer that the variable is just a path, not a file
handle.

chrismytton commented:
@tmtmtmtm 👀

tmtmtmtm commented on lib/open-uri-cached-archive.rb:
```diff
@@ -1,5 +1,43 @@
 
 class OpenUriCachedArchive
-  # Your code goes here...
+  class CachedResponse
```
I think we should be a little stricter about [one class per file](https://github.com/bbatsov/ruby-style-guide#one-class-per-file).

(I can't find a Rubocop rule that would enforce that, which I'm assuming is just me being blind, as it would be odd for them not to have something that simple…)

chrismytton replied:
Ah yep, I'm very guilty of not doing that! Moved in https://github.com/everypolitician/open-uri-cached-archive/pull/3/commits/46d6f95bbe4b1f05f7d87287aa0070bc5422548a

tmtmtmtm commented on lib/open-uri-cached-archive.rb:
```diff
@@ -1,5 +1,43 @@
+      @status = meta.delete(:status)
+      @base_uri = meta.delete(:base_uri)
+      meta.delete(:content_type)
+    end
```
There's quite a lot going on here for an `initialize` method. Perhaps some of these could become private methods instead? (`body` seems like a particularly good candidate for that)

It's also not entirely obvious what's going on with deleting things from meta. I suspect that this could be explained nicely simply by hiding it behind a method (or two) with a more intention-revealing name, but it's hard to say as I don't really understand deeply enough what's going on at the moment. This might be overkill, as this is already fairly simple, and unlikely to change much in future, so I wouldn't spend much time thinking about it if there isn't an immediately apparent quick and easy improvement.

chrismytton replied:
Yep, good shout, I've pushed some commits to refactor things into methods. I've also changed the way I'm dealing with `meta`, and added a comment to the bit that was especially odd! Hopefully that's clearer now.

tmtmtmtm commented on lib/open-uri-cached-archive.rb:
```diff
@@ -1,5 +1,43 @@
+  end
+
+  def responses
+    Dir.glob(File.join(path, '**', '*.meta')).map do |meta_file_path|
```
If you were to refactor the `body` method out above, then switching to `Pathname.glob` instead of `Dir.glob` here _might_ also help you simplify that. (Though I'm not 100% convinced, because I'm not sure whether there's a super-simple way to go from 'foo/bar.meta' to 'foo/bar' in one shot with a `Pathname`)

chrismytton replied:
I didn't even know about `Pathname.glob`, what a handy method! Turns out there is a `Pathname#sub_ext` method, which will replace the extension of the current pathname. Passing that an empty string gives us the desired behaviour. Done in https://github.com/everypolitician/open-uri-cached-archive/pull/3/commits/be62623619fde7cfb5c59a2ced51693a347cb0c0

tmtmtmtm reviewed (commented):
Generally 👍 , though a few cosmetic changes might help anyone needing to work with this again later.

chrismytton committed:
Move CachedResponse class into its own file

chrismytton committed:
Use Pathname.glob instead of Dir.glob

This will allow us to tidy up some of the code in CachedResponse.

chrismytton committed:
Make sure :status and :base_uri aren't added as headers

These keys are added by open-uri-cached, they shouldn't be returned as
part of the headers for the response.

chrismytton committed:
Use Pathname methods to extract body

chrismytton committed:
Read status and base_uri directly from meta

There's no need to set these up in the constructor.

chrismytton committed:
Move meta into its own method

chrismytton committed:
Extract CachedResponse body attribute into a method

chrismytton commented:
@tmtmtmtm 👀

tmtmtmtm reviewed (approved):
Yay!
